src/zarr/storage/_obstore.py

```diff
@@ -0,0 +1,419 @@
+        import obstore as obs
+
+        try:
+            if byte_range is None:
+                resp = await obs.get_async(self.store, key)
+                return prototype.buffer.from_bytes(await resp.bytes_async())  # type: ignore[arg-type]
+            elif isinstance(byte_range, RangeByteRequest):
+                bytes = await obs.get_range_async(
+                    self.store, key, start=byte_range.start, end=byte_range.end
+                )
+                return prototype.buffer.from_bytes(bytes)  # type: ignore[arg-type]
+            elif isinstance(byte_range, OffsetByteRequest):
+                resp = await obs.get_async(
+                    self.store, key, options={"range": {"offset": byte_range.offset}}
+                )
+                return prototype.buffer.from_bytes(await resp.bytes_async())  # type: ignore[arg-type]
+            elif isinstance(byte_range, SuffixByteRequest):
+                resp = await obs.get_async(
+                    self.store, key, options={"range": {"suffix": byte_range.suffix}}
+                )
+                return prototype.buffer.from_bytes(await resp.bytes_async())  # type: ignore[arg-type]
+            else:
+                raise ValueError(f"Unexpected byte_range, got {byte_range}")
+        except _ALLOWED_EXCEPTIONS:
+            return None
+
+    async def get_partial_values(
+        self,
+        prototype: BufferPrototype,
+        key_ranges: Iterable[tuple[str, ByteRequest | None]],
+    ) -> list[Buffer | None]:
+        # docstring inherited
+        return await _get_partial_values(self.store, prototype=prototype, key_ranges=key_ranges)
+
+    async def exists(self, key: str) -> bool:
+        # docstring inherited
+        import obstore as obs
+
+        try:
+            await obs.head_async(self.store, key)
+        except FileNotFoundError:
+            return False
+        else:
+            return True
+
+    @property
+    def supports_writes(self) -> bool:
+        # docstring inherited
+        return True
+
+    async def set(self, key: str, value: Buffer) -> None:
+        # docstring inherited
+        import obstore as obs
+
+        self._check_writable()
+        if not isinstance(value, Buffer):
+            raise TypeError(
+                f"ObjectStore.set(): `value` must be a Buffer instance. Got an instance of {type(value)} instead."
+            )
+        buf = value.to_bytes()
+        await obs.put_async(self.store, key, buf)
+
+    async def set_if_not_exists(self, key: str, value: Buffer) -> None:
+        # docstring inherited
+        import obstore as obs
+
+        self._check_writable()
+        buf = value.to_bytes()
+        with contextlib.suppress(obs.exceptions.AlreadyExistsError):
+            await obs.put_async(self.store, key, buf, mode="create")
+
+    @property
+    def supports_deletes(self) -> bool:
+        # docstring inherited
+        return True
+
+    async def delete(self, key: str) -> None:
```

Comment:
Is there an obstore method we can wrap for `delete_dir` too? That gets called somewhat frequently, so it could be a nice but small improvement.

Reply:
No, there's no specific `delete_dir` method.